File: swe/common/2023/gh-issues/gh-issues.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"

	"github.com/go-resty/resty/v2"
	"github.com/urfave/cli/v2"
)

// go run gh-issues.go aws/karpenter --count 15 | jq

type Issue struct {
	URL       string `json:"html_url"`
	Title     string `json:"title"`
	Date      string `json:"created_at"`
	Reactions struct {
		TotalCount int `json:"total_count"`
	} `json:"reactions"`
}

func fetchGitHubIssues(repoPath string, state string, count int) ([]Issue, error) {
	client := resty.New()

	url := fmt.Sprintf("https://api.github.com/repos/%s/issues", repoPath)
	params := map[string]string{
		"state":    state,
		"page":     "1",
		"per_page": "100",
	}

	var allIssues []Issue

	for {
		response, err := client.R().
			SetQueryParams(params).
			Get(url)

		if err != nil {
			return nil, err
		}

		if response.StatusCode() != 200 {
			return nil, fmt.Errorf("GitHub API request failed with status code: %d", response.StatusCode())
		}

		var issues []Issue
		err = json.Unmarshal(response.Body(), &issues)
		if err != nil {
			return nil, err
		}

		if len(issues) == 0 {
			break
		}

		allIssues = append(allIssues, issues...)
		page, _ := strconv.Atoi(params["page"])
		params["page"] = strconv.Itoa(page + 1)
	}

	sort.Slice(allIssues, func(i, j int) bool {
		return allIssues[i].Reactions.TotalCount > allIssues[j].Reactions.TotalCount
	})

	if len(allIssues) > count {
		allIssues = allIssues[:count]
	}

	return allIssues, nil
}


func main() {
	app := &cli.App{
		Name:  "gh-issues",
		Usage: "Fetch issues from a GitHub repository.",
		Flags: []cli.Flag{
			&cli.StringFlag{
				Name:    "state",
				Aliases: []string{"s"},
				Value:   "open",
				Usage:   "State of the issue (open/closed)",
			},
			&cli.IntFlag{
				Name:    "count",
				Aliases: []string{"c"},
				Value:   5,
				Usage:   "Number of issues to fetch",
			},
		},
		Action: func(c *cli.Context) error {
			repoPath := c.Args().Get(0)
			state := c.String("state")
			count := c.Int("count")

			issues, err := fetchGitHubIssues(repoPath, state, count)
			if err != nil {
				log.Fatal(err)
			}

			jsonData, err := json.MarshalIndent(issues, "", "    ")
			if err != nil {
				log.Fatal(err)
			}

			fmt.Println(string(jsonData))

			return nil
		},
	}

	err := app.Run(os.Args)
	if err != nil {
		log.Fatal(err)
	}
}

File: swe/common/2023/gh-issues/gh-issues.py
import argparse
import logging
import json

import aiohttp
import asyncio

# py gh-issues.py aws/karpenter --count 15 | jq

async def gh_issue_worker(
    repo_path: str,
    state: str,
    count: int,
):
    async with aiohttp.ClientSession() as session:
        url = f"https://api.github.com/repos/{repo_path}/issues"
        headers = {}
        params = {
            "state": state,
            "page": 1,  # Начинаем с первой страницы
            "per_page": 100,  # Указываем количество результатов на странице (максимум 100)
        }

        all_issues = []

        while True:
            async with session.get(url, headers=headers, params=params) as response:
                if response.status == 200:
                    data = await response.json()
                    if not data:
                        break  # Выход из цикла, если больше нет данных

                    bucket = []
                    for issue in data:
                        row = {
                            "url": issue.get("html_url"),
                            "title": issue.get("title"),
                            "date": issue.get("created_at"),
                            "reactions_count": issue.get("reactions").get("total_count"),
                        }
                        bucket.append(row)
                    all_issues.extend(bucket)
                    params["page"] += 1  # Увеличиваем номер страницы для следующего запроса
                else:
                    logging.warning("something goes wrong", response)
                    return None

        # Сортируем по числу лайков
        all_issues = sorted(all_issues, key=lambda x: x["reactions_count"], reverse=True)
        return all_issues[:count]


def parse_args():
    parser = argparse.ArgumentParser(description="Fetch issues from a GitHub repository.")
    parser.add_argument("repo_path", help="GitHub repository namespace")
    parser.add_argument("-s", "--state", help="State of the issue", default="open", choices=["open", "closed"])
    parser.add_argument("-c", "--count", help="GitHub API token", default=5)
    return parser.parse_args()


async def main():
    args = parse_args()

    issues = await gh_issue_worker(args.repo_path, args.state, int(args.count))
    if issues:
        print(json.dumps(issues, indent=4))
    else:
        print("Failed to fetch issues.")


if __name__ == "__main__":
    asyncio.run(main())
